### PR TITLE
Implement CN-GPU Proof-of-Work Algo

### PR DESCRIPTION
**Pull request by fireice-uk**
@psychocrypt Thanks for the hard and stressful work on this one.

### CODE REVIEW

**Comment by fireice-uk** [comment]
@brandonlehmann I had to move turtle's aglo number to 14 - please verify that this branch does not break it.

File: xmrstak/backend/amd/amd_gpu/opencl/cryptonight_gpu.cl [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+	for(size_t i = 0; i < ITERATIONS; i++)
+	{
+		mem_fence(CLK_LOCAL_MEM_FENCE);
+		((__local int*)smem)[tid] = ((__global int*)scratchpad_ptr(s, tid/4, lpad))[tid%4];
+		mem_fence(CLK_LOCAL_MEM_FENCE);
+
+		float4 rc = vs;
+
+		{
+			single_comupte_wrap(
```
Comment (Spudz76) [inline xmrstak/backend/amd/amd_gpu/opencl/cryptonight_gpu.cl]: comupte -> compute ??

**Comment by psychocrypt** [comment]
you are right I saw it during the development but missed it than. We will
fix it in a follow up since it is only naming and copy pasted

**Comment by brandonlehmann** [comment]
@fireice-uk 

Now receiving error at start-up:

```
[2019-01-30 10:27:30] : Error CL_BUILD_PROGRAM_FAILURE when calling clBuildProgram.
Build log:
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:1664:72: error: use of undeclared identifier 'lpad'; did you mean 'spad'?
                ((__local int*)smem)[tid] = ((__global int*)scratchpad_ptr(s, tid/4, lpad))[tid%4];
                                                                                     ^~~~
                                                                                     spad
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:1632:74: note: 'spad' declared here
__kernel void JOIN(cn1_cn_gpu,ALGO)(__global int *lpad_in, __global int *spad, uint numThreads)
                                                                         ^
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:1689:31: error: use of undeclared identifier 'lpad'; did you mean 'spad'?
                        *scratchpad_ptr(s , tid/4, lpad) = _mm_xor_si128(smem[tid/4], out);
                                                   ^~~~
                                                   spad
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:1632:74: note: 'spad' declared here
__kernel void JOIN(cn1_cn_gpu,ALGO)(__global int *lpad_in, __global int *spad, uint numThreads)
                                                                         ^
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:2071:12: error: use of undeclared identifier 'RCP_C'
                RCP[i] = RCP_C[i];
                         ^
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:2207:22: error: implicit declaration of function 'fast_div_v2' is invalid in C99
                        division_result = fast_div_v2(RCP, c[1], d);
                                          ^
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:2213:18: error: implicit declaration of function 'fast_sqrt_v2' is invalid in C99
                        sqrt_result = fast_sqrt_v2(c[0] + as_ulong(division_result));
                                      ^
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:2213:18: note: did you mean 'fast_div_v2'?
C:\Users\Brandon\AppData\Local\Temp\\OCL22496T1.cl:2207:22: note: 'fast_div_v2' declared here
                        division_result = fast_div_v2(RCP, c[1], d);
                                          ^
5 errors generated.

error: Clang front-end compilation failed!
Frontend phase failed compilation.
Error: Compiling CL to IR

[2019-01-30 10:27:30] : WARNING: AMD device not found
```

**Comment by psychocrypt** [comment]
thx for the report. I will fix it in 1-2hours.

**Comment by MoneroCrusher** [comment]
Who's the target demographic for this algo? CN heavy coins by any chance?

**Comment by fireice-uk** [comment]
> Who's the target demographic for this algo? CN heavy coins by any chance?

We are implementing it in Ryo.

**Comment by MoneroCrusher** [comment]
Do you have a benchmarking software? Would like to test it with Vega 56/64, RX 550 and RX 560

**Comment by fireice-uk** [comment]
Yes, you can test it on Ryo testnet -> tnpool.ryo-currency.com

**Comment by MoneroCrusher** [comment]
Great thanks. Are there any benchmark numbers available too?

**Comment by fireice-uk** [comment]
I would expect around 2000H/s on Vega64.

**Comment by MoneroCrusher** [comment]
Are there any ways to reduce power a bit and still be reasonably FPGA resistant? +75% is a lot

**Comment by fireice-uk** [comment]
> +75% is a lot

It is a worst case scenario.

> Are there any ways to reduce power a bit and still be reasonably FPGA resistant?

Not in the real world (some people are still in denial about cnv8 and FPGAs), at the end of the day it is much better for you to use 400W instead of 250W and have the FPGAs off the network than trying to compete against them.

**Comment by LeadManPL** [comment]
> Not in the real world (some people are still in denial about cnv8 and FPGAs), at the end of the day it is much better for you to use 400W instead of 250W and have the FPGAs off the network than trying to compete against them.

@fireice-uk : now, that seems reasonable. At first I was bit angry when read about that fork draining much more power and generating more heat. And I was angry because I wasn't able to find any informative page why this fork is applied in first attempt.

I assume Github is probably not ment for such discussions, thus my question: 
* Where should one search for "news" like this one in future?
* Where do you guys discuss future development direction of RYO currency?

**Pull request merged**